[Merged by Bors] - feat(linear_algebra/cayley_hamilton): the Cayley-Hamilton theorem

kim-em opened the pull request:
The Cayley-Hamilton theorem, following the proof at http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf.



---
<!-- put comments you want to keep out of the PR commit here -->

~~This PR is currently stacked on top of #3275, but the only new file here is [`linear_algebra/cayley_hamilton.lean`](https://github.com/leanprover-community/mathlib/pull/3276/files#diff-154d86bde7b844faa5021172210f55d6) if you want to see what it looks like.~~

~~Now blocked by #3352, arising from review suggestions.~~

kim-em committed:
feat(ring_theory/polynomial_algebra): polynomial A ≃ₐ[R] (A ⊗[R] polynomial R)

kim-em committed:
feat(linear_algebra/cayley_hamilton): the Cayley-Hamilton theorem

bryangingechen commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,95 @@
+import linear_algebra.nonsingular_inverse
+
+/-!
+The Cayley-Hamilton theorem, over a commutative ring.
```
```suggestion
# The Cayley-Hamilton theorem, over a commutative ring.
```

bryangingechen commented on src/ring_theory/polynomial_algebra.lean:
```diff
@@ -0,0 +1,342 @@
+import ring_theory.matrix_algebra
+import data.polynomial
+
+/-!
```
```suggestion
/-!
# Algebra isomorphism between matrices of polynomials and polynomials of matrices

```

kim-em committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

kim-em committed:
Update src/ring_theory/polynomial_algebra.lean

kim-em committed:
shorter names

kim-em committed:
Merge branch 'polynomial_algebra' of github.com:leanprover-community/mathlib into polynomial_algebra

kim-em committed:
Merge branch 'polynomial_algebra' into cayley_hamilton_squashed

kim-em committed:
Merge branch 'cayley_hamilton_squashed' of github.com:leanprover-community/mathlib into cayley_hamilton_squashed

kim-em committed:
use short names

awainverse commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,95 @@
+def char_matrix (M : matrix n n R) : matrix n n (polynomial R) :=
+matrix.scalar n (X : polynomial R) - (λ i j, C (M i j))
+
+@[simp] lemma char_matrix_apply_eq (M : matrix n n R) (i : n) :
```
The similar-looking lemmas in data.matrix.basic are named with "val". I don't know which convention lines up better with the rest of mathlib.
```suggestion
@[simp] lemma char_matrix_val_eq (M : matrix n n R) (i : n) :
```

kim-em replied:
I'm pretty sure `apply` is what's used everywhere else. I'll make a separate PR to fix `data.matrix.basic`.

kim-em replied:
#3297

kim-em replied:
I'm going to mark this as resolved --- it turns out we're already not consistent, and #3297 can be the "tracking issue" for sorting this out.

kim-em committed:
merge origin/master

kim-em committed:
fix

jcommelin commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,96 @@
+end
+
+/--
+The characteristic polynomial of a matrix `M` is given by $det (t I - M)$.
```
```suggestion
The characteristic polynomial of a matrix `M` is given by $\det (t I - M)$.
```
Alternatively just use backticks (and an `X`?)

jcommelin commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,96 @@
+  have h := calc
+    (char_poly M) • (1 : matrix n n (polynomial R))
+         = (char_matrix M).det • (1 : matrix n n (polynomial R)) : rfl
+     ... = adjugate (char_matrix M) * (char_matrix M)            : (adjugate_mul _).symm,
```
I guess we can skip the `calc` and the step in the middle, right?
```suggestion
  have h : (char_poly M) • (1 : matrix n n (polynomial R)) =
    adjugate (char_matrix M) * (char_matrix M) :=
    (adjugate_mul _).symm,
```

jcommelin replied:
Or would you prefer to keep it as it is, for educational purposes?

ChrisHughes24 commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,96 @@
+The determinant of this matrix is the characteristic polynomial.
+-/
+def char_matrix (M : matrix n n R) : matrix n n (polynomial R) :=
+matrix.scalar n (X : polynomial R) - (λ i j, C (M i j))
```
Is it worth defining `matrix.map : (R →+* S) → matrix n n R →+* matrix n n S` and using this to define this function?

jcommelin replied:
`matrix.map` should definitely exist.

kim-em replied:
Okay, I've done that as #3352, and will merge that into this, and refactor.

kim-em replied:
Okay, done.

kim-em committed:
feat(data/matrix): add matrix.map and supporting lemmas

kim-em committed:
Merge branch 'matrix_map' into cayley_hamilton_squashed

kim-em committed:
finish renaming

kim-em committed:
Merge branch 'matrix_map' into cayley_hamilton_squashed

kim-em committed:
finish renaming

kim-em committed:
Merge branch 'matrix_map' into cayley_hamilton_squashed

kim-em committed:
use matrix.map

kim-em committed:
Apply suggestions from code review

Co-authored-by: Johan Commelin <johan@commelin.net>

kim-em committed:
chore(*): import tactic.basic as early as possible, and reduce imports (#3333)

As discussed on [zulip](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/import.20refactor.20and.20library_search), #3235 had the unfortunate effect of making `library_search` and `#where` and various other things unavailable in many places in mathlib.

This PR makes an effort to import `tactic.basic` as early as possible, while otherwise reducing unnecessary imports. 

1. import `tactic.basic` "as early as possible" (i.e. in any file that `tactic.basic` doesn't depend on, and which imports any tactic strictly between `tactic.core` and `tactic.basic`, just `import tactic.basic` itself
2. add `tactic.finish`, `tactic.tauto` and `tactic.norm_cast` to tactic.basic (doesn't requires adding any dependencies)
3. delete various unnecessary imports




Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

jcommelin commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,98 @@
+/-
```
Hmm... how about renaming this file `char_poly`? I think that's what people would look for. And also, we'll probably want to add more facts later: roots are eigenvalues, simp-lemma for char_poly of 1x1 matrix, etc.

kim-em replied:
Done.

jcommelin commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,98 @@
+This holds over any commutative ring.
+-/
+-- This proof follows http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
+theorem cayley_hamilton (M : matrix n n R) :
```
Should this use `eval2` or `aeval`?
Should the name be `char_poly_eval2_self`? (I know, it looks ugly...)

kim-em replied:
It used to use eval2, but then the very first step was to rewrite it back to `(_.map _).eval _`... I'll have another look.

kim-em replied:
Yeah, I'm unconvinced that it is clearer or easier with `eval2`. The proof really does use `(_.map _).eval _`, so in order to change the theorem statement I'll need to add a penultimate line `simp only [eval₂_eq_eval_map],`

kim-em replied:
I've renamed it to `char_poly_map_eval_self`.

jcommelin commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,98 @@
+import tactic.squeeze
+
+/-!
+# The Cayley-Hamilton theorem, over a commutative ring.
```
```suggestion
# Characteristic polynomials and the Cayley-Hamilton theorem

We define characteristic polynomials of matrices and prove the Cayley–Hamilton theorem over arbitrary commutative rings.

## Main definitions

* `char_poly` is the characteristic polynomial of a matrix.

## Implementation details
```

jcommelin commented on src/linear_algebra/cayley_hamilton.lean:
```diff
@@ -0,0 +1,98 @@
+/-!
+# The Cayley-Hamilton theorem, over a commutative ring.
+
+We use a nice proof from http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
```
```suggestion
We follow a nice proof from http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
```

urkud committed:
chore(logic/embedding): reuse proofs from `data.*` (#3341)

Other changes:

* rename `injective.prod` to `injective.prod_map`;
* add `surjective.prod_map`;
* redefine `sigma.map` without pattern matching;
* rename `sigma_map_injective` to `injective.sigma_map`;
* add `surjective.sigma_map`;
* add `injective.sum_map` and `surjective.sum_map`;
* rename `embedding.prod_congr` to `embedding.prod_map`;
* rename `embedding.sum_congr` to `embedding.sum_map`;
* delete `embedding.sigma_congr_right`, add more
  general `embedding.sigma_map`.

kim-em committed:
feat(ring_theory/matrix_algebra): drop commutativity assumption (#3351)

Remove the unnecessary assumption that `A` is commutative in `matrix n n A ≃ₐ[R] (A ⊗[R] matrix n n R)`.

(This didn't cause a problem for Cayley-Hamilton, but @alexjbest and Bassem Safieldeen [realised it's not necessary](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Tensor.20product.20of.20matrices).)



Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

kim-em committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

kim-em committed:
Merge remote-tracking branch 'origin/matrix_map' into cayley_hamilton_squashed

kim-em committed:
Apply suggestions from code review

Co-authored-by: Johan Commelin <johan@commelin.net>

kim-em committed:
rename file

kim-em committed:
refactor(algebra/homology): handle co/homology uniformly (#3316)

A refactor of `algebra/homology` so homology and cohomology are handled uniformly, and factor out a file `image_to_kernel_map.lean` which gains some extra lemmas (which will be useful for talking about exact sequences).



Co-authored-by: Scott Morrison <scott.morrison@gmail.com>
Co-authored-by: Johan Commelin <johan@commelin.net>

jcommelin committed:
Merge branch 'master' into cayley_hamilton_squashed

jcommelin reviewed (approved):
Thanks :tada:

bors merge

PatrickMassot commented on src/linear_algebra/char_poly.lean:
```diff
@@ -0,0 +1,107 @@
+  -- is sent to zero, because the evaluation function puts the polynomial variable
+  -- to the right of any coefficients, so everything telescopes.
+  apply_fun (λ p, p.eval M) at h,
+  rw eval_mul_X_sub_C at h,
```
Is there any reason not to write `rwa [eval_mul_X_sub_C,  mat_poly_equiv_smul_one] at h,` instead of those three lines? Is that because of the narrative in the comments?

jcommelin replied:
I think that this proof was indeed un-golfed for readability reasons.

PatrickMassot replied:
I wonder whether we should care about the fact that `mathcomp` people always point out how [their proof](https://github.com/math-comp/math-comp/blob/3728862662bd0a5b836dfa746921954604d051ec/mathcomp/algebra/mxpoly.v#L416-L422) is only three lines long.

Maybe we can add a comment below our proof saying:
```
theorem char_poly_map_eval_self (M : matrix n n R) :
  ((char_poly M).map (algebra_map R (matrix n n R))).eval M = 0 :=
begin
  have h : (mat_poly_equiv ((char_poly M) • 1)).eval M = (mat_poly_equiv  (adjugate (char_matrix M) * (char_matrix M))).eval M := congr_arg _ (congr_arg _ (adjugate_mul _).symm),
  simpa only [mat_poly_equiv.map_mul, mat_poly_equiv_char_matrix, eval_mul_X_sub_C, mat_poly_equiv_smul_one] using h,
end
```

jcommelin replied:
Sounds good to me.
How about `by simpa ... using ...`?

PatrickMassot replied:
Indeed, on Scott's screen you could use 
```
by simpa only [mat_poly_equiv.map_mul, mat_poly_equiv_char_matrix, eval_mul_X_sub_C, mat_poly_equiv_smul_one] using (congr_arg _ (congr_arg _ (adjugate_mul _).symm) : (mat_poly_equiv ((char_poly M) • 1)).eval M = (mat_poly_equiv  (adjugate (char_matrix M) * (char_matrix M))).eval M)
```
